### PR TITLE
Dockerize pgvector

### PR DESCRIPTION
**Pull request by calufa**
- Add Dockerfile
- Update readme to include instructions on how to build a docker image

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,8 @@
+FROM postgres:9.6.21-alpine
+
+RUN apk add build-base
+
+COPY . /usr/bin/pgvector
+WORKDIR /usr/bin/pgvector
+RUN make && \
+	make install
```
Comment (calufa) [inline Dockerfile:8]: I glance over this thread, https://github.com/ankane/pgvector/issues/3, but I'm not sure what would be the best set of flags here.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+FROM postgres:9.6.21-alpine
```
Comment (ankane) [inline Dockerfile:1]: Think it'd be good to base it on `postgres` (without the tag)

File: Dockerfile (hunk 1 above)
Comment (ankane) [reply Dockerfile:8]: Just pushed a change to allow for `make OPTFLAGS=""` for maximum compatibility

File: README.md [hunk 3]
```diff
@@ -12,6 +12,18 @@ Supports L2 distance, inner product, and cosine distance
 
 [![Build Status](https://github.com/ankane/pgvector/workflows/build/badge.svg?branch=master)](https://github.com/ankane/pgvector/actions)
 
+## Docker
```
Comment (ankane) [inline README.md:15]: Let's put this under the Additional Installation Methods section (added a few days ago)

**Review by ankane — COMMENTED** [review]
Hey @calufa, thanks for the PR! Left a few comments inline.

**Comment by danielsaada** [comment]
First of all, thank you for such a useful PR, that's exactly what I was looking for!

But can you guys actually build the image?
I've tried different base images and pgvector branches, but always got this error on the make step:
```bash
...
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -O2 -march=native -ftree-vectorize -fassociative-math -fno-signed-zeros -fno-trapping-math -fPIC -shared -o vector.so src/ivfbuild.o src/ivfflat.o src/ivfinsert.o src/ivfkmeans.o src/ivfscan.o src/ivfutils.o src/ivfvacuum.o src/vector.o  -L/usr/local/lib   -L/usr/lib/llvm10/lib  -L/usr/local/lib -Wl,--as-needed  
/usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: src/ivfflat.o:(.bss+0x0): multiple definition of `ivfflat_probes'; src/ivfbuild.o:(.bss+0x0): first defined here
/usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: src/ivfinsert.o:(.bss+0x0): multiple definition of `ivfflat_probes'; src/ivfbuild.o:(.bss+0x0): first defined here
/usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: src/ivfkmeans.o:(.bss+0x0): multiple definition of `ivfflat_probes'; src/ivfbuild.o:(.bss+0x0): first defined here
/usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: src/ivfscan.o:(.bss+0x0): multiple definition of `ivfflat_probes'; src/ivfbuild.o:(.bss+0x0): first defined here
/usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: src/ivfutils.o:(.bss+0x0): multiple definition of `ivfflat_probes'; src/ivfbuild.o:(.bss+0x0): first defined here
/usr/lib/gcc/x86_64-alpine-linux-musl/10.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: src/ivfvacuum.o:(.bss+0x0): multiple definition of `ivfflat_probes'; src/ivfbuild.o:(.bss+0x0): first defined here
collect2: error: ld returned 1 exit status
make: *** [/usr/local/lib/postgresql/pgxs/src/makefiles/../../src/Makefile.shlib:309: vector.so] Error 1
```

Unfortunately I don't have time to investigate what's the issue, but I managed to build the image with this Dockerfile:
```Dockerfile
FROM postgres:12-alpine

RUN apk add build-base clang llvm

COPY . /usr/bin/pgvector
WORKDIR /usr/bin/pgvector
RUN make LDFLAGS='-Wl,--allow-multiple-definition'
RUN make install
```

I know it's ugly, but at least it works, so I hope it may help someone :)

**Comment by ankane** [comment]
Just pushed an image to Docker Hub (that uses the latest `postgres` image). Thanks for driving this @calufa!

@danielsaada Will try to see what's going on there.

**Comment by ankane** [comment]
Fixed the multiple definition error and was able to get it working with this Dockerfile:

```Dockerfile
FROM postgres:12-alpine

COPY . /tmp/pgvector

RUN apk add build-base clang llvm && \
		cd /tmp/pgvector && \
		make clean && \
		make OPTFLAGS="" && \
		make install && \
		mkdir -p /usr/share/doc/pgvector && \
		cp LICENSE README.md /usr/share/doc/pgvector && \
		rm -r /tmp/pgvector && \
		apk del build-base clang llvm
```

Doing everything in a single `RUN` that deletes the build tools results in a much smaller image.

Will close this out, but feel create a new issue if there's more Docker stuff to discuss.

**Comment by calufa** [comment]
LGTM, applogies for being late for comments.

**Pull request closed without merging**